[MRG] ENH Remove ignored_features in KBinsDiscretizer

qinhanmin2014 opened the pull request:
Now we have ``ColumnTransformer``, so we don't need to support ``ignored_features`` in ``KBinsDiscretizer`` (as we've done in ``OneHotEncoder``). See:
https://github.com/scikit-learn/scikit-learn/pull/9342#issuecomment-403673383
https://github.com/scikit-learn/scikit-learn/pull/9342#issuecomment-403676903

I think this solution will work, but we still have things to consider:
(1) Is it good to use ``_transform_selected``? I think it's acceptable, since with default ``selected="all"``, it will simply validate the input with ``check_array`` and return directly.
(2) Should we support ``inverse_transform`` for encoders other than ``ordinal``? To support this, one way is to store the fitted OntHotEncoder (or simply build a new one and set ``categories_``, so that it can support ``inverse_transform`` without fitting). The other way is to borrow some code from inverse_transform of  OntHotEncoder.

qinhanmin2014 committed:
remove ignored features

qinhanmin2014 committed:
flake8

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 9f9836d4fdb321de65efcf23300c6337bd802df2 into d1e26155e193bf2edeb51a51fb4a2dad80d95a69 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-83efc8ff8dbc5de963e008ce14de6790cdd630f7)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

qinhanmin2014 commented:
ping @jnothman @TomDLT ready for review :)

jnothman commented:
What is the benefit of using _transform_selected?

jnothman commented:
I'd be fine with storing the fitted encoder to inverse_transform. But that's an enhancement to consider later

qinhanmin2014 commented:
> What is the benefit of using _transform_selected?

No apparent benefits (maybe just avoid some duplicate code)

```
Xt = _transform_selected(X, self._transform, self.dtype, copy=True, retain_order=True)
```
is the same as
```
X = check_array(X, accept_sparse='csc', copy=True, dtype=FLOAT_DTYPES)
Xt = self._transform(X)
```

jnothman commented:
The use the latter. Much simpler to read and maintain​.

We can then, if we wish, revert all the changes to _transform_selected
currently in discrete (but they're also fairly harmless, and
_transform_selected might be removed in v0.22)

jnothman commented:
Also, it may be unnecessary to keep _transform as a separate function

On 11 July 2018 at 11:04, Joel Nothman <joel.nothman@gmail.com> wrote:

> The use the latter. Much simpler to read and maintain​.
>
> We can then, if we wish, revert all the changes to _transform_selected
> currently in discrete (but they're also fairly harmless, and
> _transform_selected might be removed in v0.22)
>

qinhanmin2014 committed:
address comment

qinhanmin2014 committed:
flake8

qinhanmin2014 commented on sklearn/preprocessing/_discretization.py:
```diff
@@ -276,23 +246,25 @@ def transform(self, X):
-        Xt = _transform_selected(X, self._transform, self.dtype,
-                                 self.transformed_features_, copy=True,
-                                 retain_order=True)
+        X = check_array(X, dtype=FLOAT_DTYPES)
```
We use ``X = check_array(X, accept_sparse='csc', copy=True, dtype=FLOAT_DTYPES)`` in ``_transform_selected``.
Here, I remove ``accept_sparse`` because we don't support sparse input here. I remove ``copy`` because I don't find it useful. I can't remove ``dtype`` because the result will change.

qinhanmin2014 reviewed (commented):
ping @jnothman ready for review

jnothman commented on sklearn/preprocessing/_discretization.py:
```diff
@@ -117,16 +109,18 @@ class KBinsDiscretizer(BaseEstimator, TransformerMixin):
 
       np.concatenate([-np.inf, bin_edges_[i][1:-1], np.inf])
 
+    You can combine ``KBinsDiscretizer`` with ``ColumnTransformer`` if you
```
Probably best to use a :class: reference here

jnothman reviewed (approved):
LGTM. Thank you

qinhanmin2014 committed:
address comment

TomDLT reviewed (changes requested):
This is much cleaner, thanks!
Just a few changes necessary

qinhanmin2014 committed:
address comment

qinhanmin2014 committed:
address comment

qinhanmin2014 committed:
flake8 sorry for the noise

TomDLT reviewed (approved):
LGTM

qinhanmin2014 commented:
Thanks @TomDLT for the review. (I'm waiting for CI, not expecting to get a response so quickly :))

> check_array is already called in _validate_X_post_fit

I choose to remove _validate_X_post_fit. Firstly, this can avoid duplicate check_array in transform. Secondly, _validate_X_post_fit block us from supporting inverse_transform for encoders other than ordinal.

> Also, copy=True is necessary since X is modified inplace. This should probably be tested.

Thanks, updated with a test. (Seems that there's not such test in the common test? not 100% sure though. )

> I also realize that self.dtype is not used anymore.

Thanks, removed.

I also let KBinsDiscretizer go through the common test :)

qinhanmin2014 commented:
CIs are green. ping @jnothman for a final check or @TomDLT if you're confident enough to merge directly. Thanks :)

TomDLT commented on sklearn/preprocessing/_discretization.py:
```diff
@@ -341,22 +285,21 @@ def inverse_transform(self, Xt):
 
-        Xt = self._validate_X_post_fit(Xt)
-        trans = self.transformed_features_
+        Xt = check_array(Xt, dtype='numeric')
```
We should probably use `FLOAT_DTYPES` here, since we modify `Xt` inplace and we may want to put float in it.

TomDLT commented on sklearn/preprocessing/_discretization.py:
```diff
@@ -274,54 +240,32 @@ def transform(self, X):
 
-        bin_edges = self.bin_edges_[trans]
-        for jj in range(X.shape[1]):
+        Xt = X.copy()
```
Why no using the copy parameter of `check_array`?
It would avoid a double copy if `X` is not a float array.

This applies also for `inverse_transform`.

qinhanmin2014 committed:
address comment

qinhanmin2014 commented:
Thanks @TomDLT. Comments addressed. I need to have a more thorough understanding of our check_array :)
